### DISCUSSION TITLE
Custom diffuse models

### DISCUSSION THREAD

**Discussion by SeleDreams** [stride3d/stride#2143] [Ideas]
Right now, Stride seems to have hardcoded the diffuse models.  I feel like that's not a good idea at all since quite a few developers might want to make their own shaders to control precisely how their object will react to light. even more when the shader doc says that shaders can be applied to all material properties. "You can use custom shaders in any material attribute" https://doc.stride3d.net/latest/en/manual/graphics/effects-and-shaders/custom-shaders.html
I feel like it would be important to allow developers to use custom shaders for lighting.
I don't really know what's the opinion of the others on this level

**Comment by tebjan** [comment]
You can extend the materials by implementing the interface and mixing in your shader. The material system is highly flexible. Just look up the diffuse models in the Stride source code, copy it, and rename it. Then it will appear there in the drop-down after it's compiled.

For example: [Stride.Rendering/Rendering/Materials/MaterialDiffuseLambertModelFeature.cs](https://github.com/stride3d/stride/blob/master/sources/engine/Stride.Rendering/Rendering/Materials/MaterialDiffuseLambertModelFeature.cs)

**Reply by SeleDreams** [reply]
Is it the same thing for Post FX ? Because I tried applying the same logic but my new post fx doesn't appear in the list

**Reply by tebjan** [reply]
This only works for color transforms, for example: [Stride.Rendering/Rendering/Images/ColorTransforms/Vignetting](https://github.com/stride3d/stride/tree/master/sources/engine/Stride.Rendering/Rendering/Images/ColorTransforms/Vignetting)

The shader keys get generated by the Visual Studio plugin automatically.
